### PR TITLE
SolverManager PECS (Producer Extends, Consumer Super)

### PR DESCRIPTION
**Pull request by ge0ffrey**
As Jiri pointed out. This solves this problem:
```
// Cat extends Animal
public Cat read(Long id) {...}
public void save(Object o) {...}

SolverManager<Long, Animal> solverManager;
solverManager.solve(1L, this::read, this::save); // Now compiles
```

### CODE REVIEW

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java [hunk 1]
```diff
@@ -38,17 +38,17 @@
     private final DefaultSolverManager<Solution_, ProblemId_> solverManager;
     private final ProblemId_ problemId;
     private final Solver<Solution_> solver;
-    private final Function<ProblemId_, Solution_> problemFinder;
-    private final Consumer<Solution_> finalBestSolutionConsumer;
+    private final Function<ProblemId_, ? extends Solution_> problemFinder;
```
Comment (yurloc) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java]: Can be `? super ProblemId_`.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java [hunk 2]
```diff
@@ -38,17 +38,17 @@
...
 
     private volatile SolverStatus solverStatus;
     private Future<Solution_> future;
 
     public DefaultSolverJob(DefaultSolverManager<Solution_, ProblemId_> solverManager, ProblemId_ problemId,
             Solver<Solution_> solver,
-            Function<ProblemId_, Solution_> problemFinder,
-            Consumer<Solution_> finalBestSolutionConsumer,
+            Function<ProblemId_, ? extends Solution_> problemFinder,
+            Consumer<? super Solution_> finalBestSolutionConsumer,
```
Comment (yurloc) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java]: The next line can be `BiConsumer<? super ProblemId_, ? super Throwable> exceptionHandler`.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverManager.java [hunk 3]
```diff
@@ -76,24 +76,24 @@ private void validateSolverFactory() {
 
     @Override
     public SolverJob<Solution_, ProblemId_> solve(ProblemId_ problemId,
-            Function<ProblemId_, Solution_> problemFinder,
-            Consumer<Solution_> finalBestSolutionConsumer,
+            Function<ProblemId_, ? extends Solution_> problemFinder,
```
Comment (yurloc) [inline optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverManager.java]: Can be `? super ProblemId_`.

File: optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java (hunk 1 above)
Comment (ge0ffrey) [reply optaplanner-core/src/main/java/org/optaplanner/core/impl/solver/DefaultSolverJob.java]: Really good catches, Jiri. Those methods should indeed support accepting a Function<Object, ...> as a problemFinder for a SolverManager<Long, ...>.

**Pull request merged**